iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,538 @@
+
+        --m_size;
+
+        // Iterator to the element following the erased one, or end() if no such element exists.
+        return iterator{const_cast<forward_list<T, CAPACITY>*>(beforeToBeErasedIter.m_cList),
+                        getPointerToListNode()[l_beforeToBeRmIdx].nextIdx};
+    }
+
+    // in case list is empty or iterator reached end()
+    return end();
+}
+
+
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::push_front(const T& ref) noexcept
+{
+    return emplace_front(ref);
+}
+
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::push_front(T&& uref) noexcept
+{
+    return emplace_front(std::forward<T>(uref));
+}
+
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::pop_front() noexcept
+{
+    auto sizeBeforeErase = m_size;
+    erase_after(before_begin());
+    return ((m_size + 1) == sizeBeforeErase);
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::insert_after(const_iterator citer,
+                                                                                     const T& r_data) noexcept
+{
+    emplace_after(citer, r_data);
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::insert_after(const_iterator citer,
+                                                                                     T&& ur_data) noexcept
+{
+    emplace_after(citer, std::forward<T>(ur_data));
+}
+
+
+template <typename T, int64_t CAPACITY>
+void forward_list<T, CAPACITY>::clear()
+{
+    for (auto iter = before_begin(); begin() != end();)
+    {
+        erase_after(iter);
+        iter = before_begin();
+    }
+}
+
+/*************************/
+// private member functions
+/*************************/
+
+/*************************/
+// iterator
+template <typename T, int64_t CAPACITY>
+forward_list<T, CAPACITY>::iterator::iterator(forward_list* parent, sizeType idx) noexcept
+    : m_list(parent)
+    , m_iterListNodeIdx(idx)
+{
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator& forward_list<T, CAPACITY>::iterator::operator++() noexcept
+{
+    // get node ref from List
+    if (m_iterListNodeIdx < INVALID_INDEX)
+    {
+        listNodePointer const nodePtr = m_list->getNodePtrFromIdx(m_iterListNodeIdx);
+        m_iterListNodeIdx = nodePtr->nextIdx;
+        return *this;
+    }
+    else
+    {
+        // do not increase beyond end() position (invalid operation), but return as-is
+        errorMessage(__PRETTY_FUNCTION__, "invalid iterator or end()");
+        return *this;
+    }
+}
+
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::iterator::operator==(const iterator rhs_iter) const noexcept
+{
+    // break recursive call for iterator::operator== by casting at least one parameter to const_cast
+    return (rhs_iter == (const_iterator) * this);
```

Comment:
Never use c-style casts. Please write `return rhs_iter == static_cast<const_iterator>(*this);` 
Additionally, what you are doing here is extremely dangerous you are casting one object to an arbitrary other object which are not related. You have to create either a conversion operator so that an iterator can be converted into a const_iterator or even better you just have one iterator class.

But in the end you should never implement two different iterator classes. You should only implement one! With const overloaded methods. Otherwise you have to write a lot of methods to cast a const_iterator into an iterator.